signac/core/synced_collection.py

```diff
@@ -41,47 +43,82 @@ def __init__(self, name=None, parent=None):
                 "Illegal argument combination, one of the two arguments, "
                 "parent or name must be None, but not both.")
 
+    @classmethod
+    def __init_subclass__(cls):
+        """Add  ``_validator`` attribute to every subclass.
+
+        Subclasses contain a list of validators that are applied to collection input data.
+        Every subclass must have a separate list so that distinct sets of validators can
+        be registered to each of them.
+        """
+        cls._validators = []
+
     @classmethod
     def register(cls, *args):
-        """Register the synced data structures.
+        r"""Register the synced data structures.
 
-        Registry is used when recursively converting synced data structures to determine
-        what to convert their children into.
+        The registry is used by :meth:`from_base` to determine the appropriate subclass
+        of :class:`SyncedCollection` that should be constructed from a particular object.
+        This functionality is necessary for converting nested data structures because
+        given, for instance, a dict of lists, it must be possible to map the nested lists to
+        the appropriate subclass of :class:`SyncedList` corresponding to the desired
+        backend.
 
         Parameters
         ----------
-        *args
+        \*args
             Classes to register
         """
         if not hasattr(cls, 'registry'):
             cls.registry = defaultdict(list)
-        for _cls in args:
-            cls.registry[_cls.backend].append(_cls)
+        for base_cls in args:
+            cls.registry[base_cls._backend].append(base_cls)
+
+    @property
+    def validators(self):
```

Comment:
@bdice @vishav1771 Was there any further discussion on this? My inclination is that this may not really be safe because users could always add new validators, but we could also forbid that explicitly.

If it was just decided that this request is out of scope for this PR, that is fine with me.

Reply:
@vyasr Hmm. I guess that caching wouldn't be safe if new validators were applied to a parent class (it would be easy to empty the cache if a validator were added to _this_ class, but it would be hard to propagate that information from a parent class if the parent class's validators changed). Maybe we just leave this as it is for now. I was concerned that this might affect performance. We'll need to do thorough benchmarking/profiling before releasing the new backends. I recommend that we should defer this discussion until we look at the performance more carefully.

Reply:
Yes we'll need to profile everything once everything is feature complete. On the plus side, we've added a pretty good test suite so if we need to optimize we should be able to validate the changes relatively easy.